### PR TITLE
Cleaned up Lidar branch to merge with dev

### PR DESCRIPTION
**Pull request by zlite**
Hopefully this is clean enough to review. Rebased on dev and stripped of all commits from non-relevant branches on my side

### CODE REVIEW

**Comment by zlite** [comment]
The reason the CI tests are failing is that the Lidar code has a dependency on the rplidar library.  The instructions say to install it, but apparently the CI system here expects that to be listed in some dependency/environment install file. 

Am I supposed to add it to the repo-wide pc/linux/mac yml file even through most people won't be using lidar? Seems weird

**Comment by zlite** [comment]
Note: this now supports both the Basic and the Complete templates

File: donkeycar/parts/controller.py [hunk 1]
```diff
@@ -229,6 +228,108 @@ def set_deadzone(self, val):
         self.dead_zone = val
 
 
+class RCReceiver:
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:231]: It looks like you merged parts of the RCReceiver branch into this PR - is this intended or an accident?

File: donkeycar/parts/controller.py [hunk 2]
```diff
@@ -1527,4 +1628,4 @@ def get_js_controller(cfg):
         button, button_state, axis, axis_val = js.poll()
         if (button is not None or axis is not None):
             print(button, button_state, axis, axis_val)
-            time.sleep(0.1)
+            time.sleep(0.1)
```
Comment (DocGarbanzo) [inline donkeycar/parts/controller.py:1631]: Please keep the newline at EOF.

File: donkeycar/parts/lidar.py [hunk 3]
```diff
@@ -47,6 +80,62 @@ def shutdown(self):
...
+    https://pypi.org/project/PyLidar3/
+    '''
+    def __init__(self, port='/dev/ttyUSB0'):
+        import PyLidar3
+        self.port = port
+        self.distances = [] #a list of distance measurements 
+        self.angles = [] # a list of angles corresponding to dist meas above
+        self.lidar = PyLidar3.YdLidarX4(port)
+        if(self.lidar.Connect()):
+            print(self.lidar.GetDeviceInfo())
```
Comment (DocGarbanzo) [inline donkeycar/parts/lidar.py:94]: In general we started to move from `print` to `logging`. Not an issue here in particular, b/c we will have to move more `print` statements to be consistent in the future.

File: donkeycar/parts/tub_v2.py [hunk 4]
```diff
@@ -52,6 +52,8 @@ def write_record(self, record=None):
                     contents[key] = int(value)
                 elif input_type == 'boolean':
                     contents[key] = bool(value)
+                elif input_type == 'nparray':
+                    contents[key] = (np.array(value)).tolist()
```
Comment (DocGarbanzo) [inline donkeycar/parts/tub_v2.py:56]: If `type(value) == np.array`, then you don't need the `np.array(...)` around it.

File: donkeycar/parts/tub_v2.py [hunk 5]
```diff
@@ -130,3 +132,38 @@ def shutdown(self):
         self.close()
 
 
+
+
+class TubWiper:
```
Comment (DocGarbanzo) [inline donkeycar/parts/tub_v2.py:137]: This is part of the RCReceiver branch, pls see comment above.

File: donkeycar/parts/datastore.py [hunk 6]
```diff
@@ -235,6 +235,9 @@ def put_record(self, data):
                 name = self.make_file_name(key, ext='.png')
                 img.save(os.path.join(self.path, name))
                 json_data[key]=name
+                
+            elif typ == 'nparray':                         # if array type
+                json_data[key] = (np.array(val)).tolist()  # store the input as an array into tub
```
Comment (DocGarbanzo) [inline donkeycar/parts/datastore.py:240]: Same comment as below - if `type(val) == np.array` then we shouldn't need to wrap it into `(np.array(val))`, so `val.tolist()` should work.

File: donkeycar/templates/basic.py [hunk 7]
```diff
@@ -12,10 +13,11 @@
 from docopt import docopt
 
 import donkeycar as dk
-from donkeycar.parts.tub_v2 import TubWriter
+from donkeycar.parts.tub_v2 import TubWriter, TubWiper
 from donkeycar.parts.datastore import TubHandler
-from donkeycar.parts.controller import LocalWebController
+from donkeycar.parts.controller import LocalWebController, RCReceiver
```
Comment (DocGarbanzo) [inline donkeycar/templates/basic.py:18]: As above, these are the RC receiver changes - should we keep them separate?

File: donkeycar/templates/cfg_basic.py [hunk 8]
```diff
@@ -49,10 +50,27 @@
...
 
-#TRAINING
-# The default AI framework to use. Choose from (tensorflow|pytorch)
-DEFAULT_AI_FRAMEWORK='tensorflow'
+DRIVE_TRAIN_TYPE = "SERVO_ESC" # SERVO_ESC|DC_STEER_THROTTLE|DC_TWO_WHEEL|SERVO_HBRIDGE_PWM|PIGPIO_PWM|MM1|MOCK
+
+# #LIDAR
+USE_LIDAR = False
+LIDAR_TYPE = 'RP' #(RP|YD)
+LOWER_LIMIT = 44 # angles that will be recorded. Use this to block out obstructed areas on your car, or looking backwards. Note that for the RP A1M8 Lidar, "0" is in the direction of the motor
```
Comment (DocGarbanzo) [inline donkeycar/templates/cfg_basic.py:58]: Probably better to call these `LIDAR_LOWER_LIMIT`,  `LIDAR_UPPER_LIMIT` as otherwise it's hard to know what these attributes refer to.

File: donkeycar/tests/lidar_test.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
+from rplidar import RPLidar
+import time
+import serial
+
+import glob
+import time # Time module
+temp_list = glob.glob ('/dev/ttyUSB*')
```
Comment (DocGarbanzo) [inline donkeycar/tests/lidar_test.py:7]: How does this test work in the PC or on the car if you don't have Lidar?

File: donkeycar/templates/cfg_basic.py [hunk 10]
```diff
@@ -122,11 +136,11 @@
 #then extract that and modify DONKEY_SIM_PATH.
 DONKEY_GYM = False
 DONKEY_SIM_PATH = "path to sim" #"/home/tkramer/projects/sdsandbox/sdsim/build/DonkeySimLinux/donkey_sim.x86_64" when racing on virtual-race-league use "remote", or user "remote" when you want to start the sim manually first.
-DONKEY_GYM_ENV_NAME = "donkey-generated-track-v0" # ("donkey-generated-track-v0"|"donkey-generated-roads-v0"|"donkey-warehouse-v0"|"donkey-avc-sparkfun-v0")
+DONKEY_GYM_ENV_NAME = "donkey-mountain-track-v0" # ("donkey-generated-track-v0"|"donkey-generated-roads-v0"|"donkey-warehouse-v0"|"donkey-avc-sparkfun-v0")
```
Comment (DocGarbanzo) [inline donkeycar/templates/cfg_basic.py:139]: Is this change intended?

File: donkeycar/parts/tub_v2.py (hunk 5 above)
Comment (zlite) [reply donkeycar/parts/tub_v2.py:137]: As mentioned, I need that in. It will soon be standard so no harm in having it

File: donkeycar/templates/cfg_basic.py (hunk 10 above)
Comment (zlite) [reply donkeycar/templates/cfg_basic.py:139]: No, not intended. Must have picked it up from dev

File: donkeycar/parts/tub_v2.py (hunk 5 above)
Comment (DocGarbanzo) [reply donkeycar/parts/tub_v2.py:137]: Ok fine, just wanted your confirmation.

**Comment by DocGarbanzo** [comment]
@zlite - I have addressed all my comments in your branch (currently only locally). Are you ok for me to push it into your repo - that'll update the PR here automatically. Otherwise I'll do a PR into your fork.

**Comment by zlite** [comment]
Yes, please do push to my repo. They all look fine, but I'll pull and test just to be sure

**Comment by DocGarbanzo** [comment]
@zlite - super, done this now.

**Review by DocGarbanzo — APPROVED** [review]
Looks good to me now.
